Zebra Nexthop Group Rework and Kernel Nexthop Object API Init

sworleys opened the pull request:
These patches provide a re-work of the internal zebra nexthop group structure in line with the new kernel nexthop API: https://lore.kernel.org/netdev/20190524214308.18615-6-dsahern@kernel.org/T/

This code implements a hashing functionality for routes we receive from upper level protocols to share nexthop groups and create objects we can track along with the kernels installed ones.

The design can be referenced here:
https://docs.google.com/presentation/d/1SeDS5b-Wgmp-2T_9povfHscP6Xpaihff_xsxWdTlKDE/edit?usp=sharing (slide 7-9)


The dataplane code paths for the new kernel API nexthop creation have all been written but are currently forced off. After a bit more testing I will enable that code path with a separate PR. Hence, no nexthop objects will be installed on a kernel at the moment.

Some significant memory improvements can be seen:

300,000 routes x32 ECMP


**Master:**
```
alfred# sharp install routes 2.2.2.1 nexthop-group blue 300000                
alfred# show memory                                                                                                    
Memory statistics for zebra:                                                                                           
System allocator statistics:                                                                                           
  Total heap allocated:  > 2GB                                                                                         
  Holding block headers: 8196 KiB                                                                                      
  Used small blocks:     0 bytes                                                                                       
  Used ordinary blocks:  > 2GB                                                                                         
  Free small blocks:     2112 bytes                                                                                    
  Free ordinary blocks:  80 MiB                                                                                        
  Ordinary blocks:       315812                                                                                        
  Small blocks:          63                                                                                            
  Holding blocks:        1                                                                                             
(see system documentation for 'mallinfo' for meaning) 
```

**This Patch**
```
alfred# sharp install routes 2.2.2.1 nexthop-group blue 300000
alfred# show memory                                                                                                    
Memory statistics for zebra:                                                                                           
System allocator statistics:                                                                                           
  Total heap allocated:  204 MiB                                                                                       
  Holding block headers: 8196 KiB                                                                                      
  Used small blocks:     0 bytes                                                                                       
  Used ordinary blocks:  130 MiB                                                                                       
  Free small blocks:     2144 bytes                                                                                    
  Free ordinary blocks:  74 MiB                                                                                        
  Ordinary blocks:       250592                                                                                        
  Small blocks:          64                                                                                            
  Holding blocks:        1                                                                                             
(see system documentation for 'mallinfo' for meaning)
```


**New commands**
`show nexthop-group [ip|ipv6] [vrf] [id]`

- Display currently created nexthop groups and whether they are valid/installed in the kernel

`show interface nexthop-group [vrf] `

- Nexthop groups pointing to this interface

`show ip route [prefix] nexthop-group`

- Ip route info along with nexthop group ID.


**TODO**

- [x] Add some docs to the guide

donaldsharp commented:
@louberger -> any input on what has gone wrong here?

louberger commented:
MAKEINFO doc/user/_build/texinfo/frr.info zebra/zebra_vty.c: In function 
vty_show_ip_route zebra/zebra_vty.c:464:13: error: struct route_entry has 
no member named nexthop_num ; did you mean nexthop_mtu re->nexthop_num); 
^~~~~~~~~~~ nexthop_mtu zebra/zebra_vty.c:466:13: error: struct route_entry 
has no member named nexthop_active_num ; did you mean nexthop_mtu 
re->nexthop_active_num); ^~~~~~~~~~~~~~~~~~ nexthop_mtu Makefile:7127: 
recipe for target 'zebra/zebra_vty.o' failed make[1]: *** 
[zebra/zebra_vty.o] Error 1 make[1]: *** Waiting for unfinished jobs.... 
make[1]: Leaving directory 
'/home/sw/regression/frr-reference/master/github/frr.github' Makefile:4032: 
recipe for target 'all' failed make: *** [all] Error 2


----------
On August 29, 2019 11:54:41 AM LabN CI User <notifications@github.com> wrote:

> ## &#x1F6d1; Basic BGPD CI results: FAILURE
>
> <details>
> <summary>Results table</summary>
>
> _|_
> -|-
> Result | FAILURE git merge/4897 e04c35ee frr.github Build
> Date |08/29/2019
> Start |11:53:05
> Finish |11:54:05
> Run-Time |01:00
> Total |
> Pass |
> Fail |
> Valgrind-Errors |
> Valgrind-Loss |
> Details |vncregress-2019-08-29-11:53:05.txt
> Log |make-2019-08-29-11:53:05.out.bz2
> Memory |
>
> For details, please contact louberger
>
> </details>
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/FRRouting/frr/pull/4897#issuecomment-526249447

donaldsharp commented:
@louberger -> the code in the above compile failure is different then the branch that Sworley has put forward( there is nothing about nexthop_num on lines 464 or 466 of zebra_vty.c )

Has something else gone wrong?

sworleys commented:
> ```MAKEINFO doc/user/_build/texinfo/frr.info zebra/zebra_vty.c: In function vty_show_ip_route zebra/zebra_vty.c:464:13: error: struct route_entry has no member named nexthop_num ; did you mean nexthop_mtu re->nexthop_num); ^~~~~~~~~~~ nexthop_mtu zebra/zebra_vty.c:466:13: error: struct route_entry has no member named nexthop_active_num ; did you mean nexthop_mtu re->nexthop_active_num); ^~~~~~~~~~~~~~~~~~ nexthop_mtu Makefile:7127: recipe for target 'zebra/zebra_vty.o' failed make[1]: *** [zebra/zebra_vty.o] Error 1 make[1]: *** Waiting for unfinished jobs.... make[1]: Leaving directory '/home/sw/regression/frr-reference/master/github/frr.github' Makefile:4032: recipe for target 'all' failed make: *** [all] Error 2 ```
> […](#)
> ----------
> On August 29, 2019 11:54:41 AM LabN CI User ***@***.***> wrote: ## &#x1F6d1; Basic BGPD CI results: FAILURE <details> <summary>Results table</summary> _|_ -|- Result | FAILURE git merge/4897 [e04c35e](https://github.com/FRRouting/frr/commit/e04c35eee459194cb16ba55b93a11092b6daca84) frr.github Build Date |08/29/2019 Start |11:53:05 Finish |11:54:05 Run-Time |01:00 Total | Pass | Fail | Valgrind-Errors | Valgrind-Loss | Details |vncregress-2019-08-29-11:53:05.txt Log |make-2019-08-29-11:53:05.out.bz2 Memory | For details, please contact louberger </details> -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#4897 (comment)](https://github.com/FRRouting/frr/pull/4897#issuecomment-526249447)

Had to re-base to expose these.

Fixed them to use the new API and force-pushed.

sworleys commented:
@mjstapp So looks like the failing topotests are due to the fact that the `nexthop group` data is shared between a route that is installed and one that isn't.

Since one is installed, it is getting the `fib` flag on it and then when compare that nexthop while looking at the route that isn't installed we see that nexhtop has the fib flag and mark it as failed. In the old style this would make sense since the data isn't shared.

I will figure out a proper fix for this test and add it to the PR.

sworleys commented:
ci:rerun

sworleys commented on lib/nexthop_group.c:
```diff
@@ -112,6 +206,13 @@ void nexthop_group_delete(struct nexthop_group **nhg)
 	XFREE(MTYPE_NEXTHOP_GROUP, *nhg);
 }
 
+void nexthop_group_free_delete(struct nexthop_group **nhg)
```
When I wrote this there was somewhere in our code base that had a use-case to delete the group but not the nexthops in it. But I don't see that anymore so yes I think we can reduce it.

sworleys commented on zebra/zebra_router.c:
```diff
@@ -253,4 +263,22 @@ void zebra_router_init(void)
+ */
+void zebra_router_cleanup(void)
+{
+	zebra_nhg_cleanup_tables(zrouter.nhgs_id);
```
I was pretty much following the same style used in `zebra_router_terminate()` where it calls the appropriate `mlag` and `vxlan` files api's.

But that function is just boilerplate right now if you look at the code. The nhg's should be removed from the kernel via `route delete` as the refcnt's drop. I think we can probably just remove that function altogether for the time being so I will do that.

sworleys commented on zebra/zebra_nhg.h:
```diff
@@ -24,6 +24,245 @@
+
+PREDECL_RBTREE_UNIQ(nhg_connected_tree);
+
+struct nhg_hash_entry {
```
In the top of zebra router it has this:

```
/*                                                                               
 * This header file contains the idea of a router and as such                    
 * owns data that is associated with a router from zebra's                       
 * perspective.                                                                  
 */                                                                              
```

Since that is where the route tables are kept as well I felt its the appropriate place for nexthop group hashes.

The rb tree you are referencing here are internal ones only used from `nhg -> nhg` via the depend/dependent relationship.

But yes, I think a comment is probably owed there.

sworleys commented on zebra/zebra_nhg.h:
```diff
@@ -24,6 +24,245 @@
+
+
+void zebra_nhg_init(void);
+void zebra_nhg_terminate(void);
```
Yep those need to be removed.

sworleys commented:
When we rebase, make sure we don't miss #4972

sworleys commented:
Re-based and force-pushed

sworleys commented:
```
<stdin>:40: new blank line at EOF.
+
warning: 1 line adds whitespace errors.
Report for kernel_netlink.c | 2 issues
===============================================
< WARNING: line over 80 characters
< #898: FILE: /tmp/f1-31624/kernel_netlink.c:898:
Report for rt_netlink.c | 4 issues
===============================================
< WARNING: line over 80 characters
< #458: FILE: /tmp/f1-31624/rt_netlink.c:458:
< WARNING: line over 80 characters
< #466: FILE: /tmp/f1-31624/rt_netlink.c:466:
Report for zebra_errors.c | 4 issues
===============================================
< WARNING: line over 80 characters
< #317: FILE: /tmp/f1-31624/zebra_errors.c:317:
< WARNING: line over 80 characters
< #607: FILE: /tmp/f1-31624/zebra_errors.c:607:
```

The blank line seems to be false negative (i cant find it).
The line limit ones are save to ignore as they are either strings or were like that before my patch.

riw777 commented on zebra/zebra_nhg.c:
```diff
@@ -256,7 +256,7 @@ struct nhg_hash_entry *zebra_nhg_find(struct nexthop_group *nhg,
 			flog_warn(
 				EC_ZEBRA_DUPLICATE_NHG_MESSAGE,
-				"Nexthop Group from kernel with ID (%d) is a duplicate, ignoring",
+				"Nexthop Group from with ID (%d) is a duplicate, ignoring",
```
Probably want to remove the "from" as well in this change.

sworleys replied:
This has already been addressed:
```
        flog_warn(EC_ZEBRA_DUPLICATE_NHG_MESSAGE,                                
                  "Nexthop Group with ID (%d) is a duplicate, ignoring",         
                  nhe->id);    
```
in `zebra_nhg_set_dup()`

I believe I addressed all of @mjstapp 's comments as well.

sworleys commented:
Force pushed to rebase.

mjstapp commented on zebra/zebra_nhg.h:
```diff
@@ -24,6 +24,245 @@
+ * It is designed to mimic the netlink nexthop_grp
+ * struct in include/linux/nexthop.h
+ */
+struct nh_grp {
```
why? why not add ... a weight to the main group struct and do the conversion from netlink to os-neutral?

sworleys replied:
I am not following, this is an OS neutral one right? I needed this to be able to store a information about a group we received from the kernel and enque it onto the rib metaqueue for processing later.

I did this to account for the future dataplane enhancements so all routes from the kernel are received in the other ptrhead as well. Rather than creating nhg_hash_entry objects in the netlink code and adding them directly from there.

I don't think this belongs as part of the main group struct since its really only for queuing a group to be handled later.

mjstapp commented on zebra/zebra_nhg.h:
```diff
@@ -24,6 +24,245 @@
+ * Hashtables contiaining entries found in `zebra_router`.
+ */
+
+struct nhg_hash_entry {
```
we'll have nh_grp, nexthop_grp, nhg_ctx, nhg_hash_entry, and nexthop_group structs - that's a bit confusing. can we keep the netlink-specific things in a netlink-specific place (or eliminate them)?  could we extend the existing nexthop_group struct to include the extra things that have been put into `nhg_hash_entry`? the special "this is not a group" case in `nhg_hash_entry` - couldn't we eliminate that by always having a real nexthop, even if there's just one?

sworleys replied:
> we'll have nh_grp, nexthop_grp, nhg_ctx, nhg_hash_entry, and nexthop_group structs - that's a bit confusing. can we keep the netlink-specific things in a netlink-specific place (or eliminate them)? 
See my other reply for info on why the `nh_grp` array. But I believe the netlink specific stuff is separate. I re-used a similar struct as it yes, but that was intentional because I wanted to have the notion of arrays of IDs/Weights inside the `nhg_ctx` object without having to allocate a nexthop struct for all of them.

> could we extend the existing nexthop_group struct to include the extra things that have been put into `nhg_hash_entry`? 
That is a lot of extra memory to be putting into a lib struct that the other daemons will not be using for a while though. I assume this is why @donaldsharp went with the `wrapper` approach rather than putting it directly into a lib struct.

> the special "this is not a group" case in `nhg_hash_entry` - couldn't we eliminate that by always having a real nexthop, even if there's just one?

I am not sure what special case you are referring to. There is always a real `nexthop` and `nexthop_group` contained in the `nhg_hash_entry`

mjstapp commented on zebra/zebra_rib.c:
```diff
@@ -2412,7 +2424,13 @@ void rib_unlink(struct route_node *rn, struct route_entry *re)
+	} else if (re->ng)
+		nexthop_group_delete(&re->ng);
+
 	nexthops_free(re->fib_ng.nexthop);
```
are we not using the 'fib nexthop-group' concept? we did add that to try to deal with a couple of problems

sworleys replied:
That is still retained, its just not created as a nhg_hash_entry.

When we enable the kernel nexthop group API changes will need to be made there though. Still figuring that one out. They might not even be useful at that point to be honest.

mjstapp commented on zebra/zebra_nhg.c:
```diff
@@ -26,14 +26,1088 @@
+		ret = nhg_ctx_process_new(ctx);
+		if (nhg_ctx_get_count(ctx) && ret == -ENOENT
+		    && nhg_ctx_get_status(ctx) != NHG_CTX_REQUEUED) {
+			/* Depends probably came before group, re-queue.
```
Don't follow what's happening here - what does "depends ... came before group" mean? A block of comments about what the flow is here between the kernel/OS and zebra would be really helpful.

sworleys replied:
This is to account for the possibility of processing a group before any of its depends have been handled. We re-queue once and only once.

Agreed, it is worded terribly, I will add sufficient commenting.

mjstapp commented on zebra/zebra_nhg.c:
```diff
@@ -26,14 +26,1088 @@
+		zebra_nhg_handle_uninstall(nhe);
+}
+
+static int nhg_ctx_process_new(struct nhg_ctx *ctx)
```
We don't flag the KERNEL groups, is that right? why not - wouldn't that help, as we track routes that are KERNEL/SYSTEM ?

sworleys replied:
we get that info from the `nhe->type` field, which comes in as the protocol field from the kernel. If its not RTPROT_ZEBRA, we don't consider it ours. We do it similarly with routes.

sworleys commented on lib/nexthop.c:
```diff
@@ -364,6 +364,19 @@ struct nexthop *nexthop_next(struct nexthop *nexthop)
 }
 
+/* Return the next nexthop in the tree that is resolved and active */
+struct nexthop *nexthop_next_active_resolved(struct nexthop *nexthop)
```
probably, will do

sworleys commented on lib/nexthop.c:
```diff
@@ -364,6 +364,19 @@ struct nexthop *nexthop_next(struct nexthop *nexthop)
+	return next;
+}
+
 unsigned int nexthop_level(struct nexthop *nexthop)
```
will do

sworleys commented on lib/nexthop_group.c:
```diff
@@ -232,6 +329,35 @@ uint32_t nexthop_group_hash(const struct nexthop_group *nhg)
+	return key;
+}
+
+void nexthop_group_mark_duplicates(struct nexthop_group *nhg)
```
A fair point. They were in the code before so I elected to keep them at least for now to avoid unforeseen errors. But I think we could get rid of them in a future PR to optimize. I don't think it hurts much to leave them in for now.

sworleys commented on zebra/zebra_nhg.c:
```diff
@@ -26,14 +26,1088 @@
+	return ctx;
+}
+
+static bool zebra_nhg_contains_dup(struct nhg_hash_entry *nhe)
```
Ah I see there is some confusion here. I had forgotten there existed a similar flag on the nexthop itself. This flag is to indicate the kernel sent us this nexthop group and we had one already that was identical, so we are only tracking it for its ID.

I will change it to be something better.

sworleys commented on zebra/zebra_nhg.c:
```diff
@@ -26,14 +26,1088 @@
+		struct nhg_connected *rb_node_dep = NULL;
+
+		/* If anthing else in the group is valid, the group is valid */
+		frr_each(nhg_connected_tree, &nhe->nhg_dependents,
```
Good catch, will fix.

sworleys commented:
@mjstapp Addressed your changes, added some of my own, rebased and force pushed.

New commits start at https://github.com/FRRouting/frr/pull/4897/commits/34132e3c0bac622a2719bc7711636e4dd553c4d6

sworleys commented:
> ### Warnings Generated during build:
> Checkout code: Successful with additional warnings
> ```
> <stdin>:40: new blank line at EOF.
> +
> warning: 1 line adds whitespace errors.

Idk what the new blank line is referring to, I can't find it.

> Report for kernel_netlink.c | 2 issues
> ===============================================
> < WARNING: line over 80 characters
> < #898: FILE: /tmp/f1-17295/kernel_netlink.c:898:

This is a log string so its intentional.

> Report for rt_netlink.c | 6 issues
> ===============================================
> < WARNING: line over 80 characters
> < #458: FILE: /tmp/f1-17295/rt_netlink.c:458:
> < WARNING: line over 80 characters
> < #466: FILE: /tmp/f1-17295/rt_netlink.c:466:

> Report for zebra_errors.c | 4 issues
> ===============================================
> < WARNING: line over 80 characters
> < #317: FILE: /tmp/f1-17295/zebra_errors.c:317:

Log string, intentional.

> < WARNING: line over 80 characters
> < #607: FILE: /tmp/f1-17295/zebra_errors.c:607:

Log string, intentional.


The others I fixed, re-based, and force-pushed.

piotrjurkiewicz commented:
Since this pull request, zebra does not push routes to dataplane (kernel of FPM) anymore. Routes are computed correctly by routing protocol, but they are marked as inactive and not pushed downwards.

I am unable to bisect into the exact commit inside this pull request which introduced the bug because of compilation errors.

```
r0> show ip route

Codes: K - kernel route, C - connected, S - static, R - RIP,
       O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
       F - PBR, f - OpenFabric,
       > - selected route, * - FIB route, q - queued route, r - rejected route

O   10.127.0.0/30 [110/10] is directly connected, eth2 inactive, 00:01:11
C>  10.127.0.0/30 is directly connected, eth2 inactive, 00:01:20
O   10.127.0.4/30 [110/10] is directly connected, eth3 inactive, 00:01:11
C>  10.127.0.4/30 is directly connected, eth3 inactive, 00:01:21
O   10.127.0.8/30 [110/12] via 10.127.0.22, eth1 inactive, 00:00:16
O   10.127.0.12/30 [110/11] via 10.127.0.22, eth1 inactive, 00:00:26
O   10.127.0.16/30 [110/12] via 10.127.0.22, eth1 inactive, 00:00:16
O   10.127.0.20/30 [110/1] is directly connected, eth1 inactive, 00:00:31
C>  10.127.0.20/30 is directly connected, eth1 inactive, 00:01:21
O   10.127.0.24/30 [110/2] via 10.127.0.22, eth1 inactive, 00:00:26

r0> show interface

Interface eth1 is up, line protocol is up
  Link ups:       0    last: (never)
  Link downs:     0    last: (never)
  vrf: default
  index 4 metric 0 mtu 1500 speed 10000
  flags: <UP,BROADCAST,RUNNING,MULTICAST>
  Type: Ethernet
  HWaddr: 86:5b:ae:53:f8:86
  inet 10.127.0.21/30
  inet6 fe80::845b:aeff:fe53:f886/64
  Interface Type VETH
  Parent interface: eth3
Interface eth2 is up, line protocol is up
  Link ups:       1    last: 2019/11/17 03:04:02.24
  Link downs:     0    last: (never)
  vrf: default
  index 2 metric 0 mtu 1500 speed 10000
  flags: <UP,BROADCAST,RUNNING,MULTICAST>
  Type: Ethernet
  HWaddr: 6a:45:a3:e0:0f:57
  inet 10.127.0.2/30
  inet6 fe80::6845:a3ff:fee0:f57/64
  Interface Type VETH
  Parent interface: eth2
Interface eth3 is up, line protocol is up
  Link ups:       0    last: (never)
  Link downs:     0    last: (never)
  vrf: default
  index 3 metric 0 mtu 1500 speed 10000
  flags: <UP,BROADCAST,RUNNING,MULTICAST>
  Type: Ethernet
  HWaddr: 6a:a4:74:39:ed:9a
  inet 10.127.0.6/30
  inet6 fe80::68a4:74ff:fe39:ed9a/64
  Interface Type VETH
  Parent interface: eth2
Interface lo is up, line protocol is up
  Link ups:       0    last: (never)
  Link downs:     0    last: (never)
  vrf: default
  index 1 metric 0 mtu 65536 speed 0
  flags: <UP,LOOPBACK,RUNNING>
  Type: Loopback
  Interface Type Other
```

```
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
085304dc09bad042df987d9fb62272afd993f7fe
18cf7f15d635ba9738e98b7eeb5ff50cfacbec0e
13dfea50e4aba145d57ce32247208d7108d5e803
21615102a0ef57a321999c3d1289935e4a1d95d6
3082f99c48abad6e04eaebb8b775e762bd18429a
6ccc2b28efa3223385ae764e667ed9bec1231f85
71593b3f0f67b797fa90571fa7c5a56cd5d8ef89
7192bb23fba993b072c5e3aabaf5ad6eae523fef
8f8d9845c6bc283e5a9f7339f9d3f2884774589b
e22e80010ee4226cbb4c1cb570d520ff071add89
a6e6a6d825f98ecd3c074dfa473aec2d5da4cdb1
ddaee0c7d3ca75040a9b28289a8d961084b000ea
7b683a96e4227956cfe6236628885b1421b62781
de3f54881950c999ba5034bf525783a2c9d8e931
3230a4dba5f3ee7dfe0d8d12f23b212c62215faa
bbb3940ed11d62df8263bb97ef393c1d7de0220b
1c3d2890408e05f60b4864a93a1dff0c23ab346c
cb50cbc96ee0b6c69e7c80d00c3d26070f4a1651
98cda54a9543ea125e5e1eea6621c453f407edb2
f862383fc9f0055dc068d5374e5efc7b49a23e2e
7f1abf7926b71ac7c7170883d0d8d0039ba269b0
fe593b781d8a927bd496c77799312cc64a565141
144a1b34dfc37fdd1174e7a2f150f2315730d8f7
a15d4c0089a6b8f59932df39357725feadfb9f91
0c8215cbab59480616fe31d8a81323070a21137b
4505578be0304dd314e96e352be13d0925584497
2f65aee06ad31d720a316f9d7dee90dcdaebcd41
7512f617c7f6e9258f0b8f53f6d556d1e36611e9
ab942eb2853310fa76e1bcdad422eed441adeb00
da1371420204efbdaef17138392b146708cfbf4c
6df591527f8de78378e24096de6e78f42797b737
86946a2207c3f2c177b478c9cf64f491a922df5c
055a3fa698319845cc6f30376db90411f6b7be24
f54ef6a5ba251ba091a4ec0079008ec7a6f81883
7fd392cc4d804aae6bd9d1452ba5b108ecdc70a2
9ddc33505af6ad1fd99cd23352fc2dc99462afbb
20822f9d2e450cebfefbb14cc70a189a52df063c
5155d86c6f8752dc9bdfeb9552f7f06c02567e17
148a0103c6176b5bb3f3dff531ed06dda754f3c7
2d3c57e671d713feb7a6ef34982c375616368dc6
1e9aad82ba65df9c50eb819334cb2caace57df6c
7286ac02a3023214c0afc3bc8415a31231dae808
3057df51bd471798f4c221531112340e11b7fe46
9834bb52aebb46fc4ba85e959cc98adda996a4b3
139ddad8f12cf6dbf2993af196cc33476c2b79cf
9ed6c34a871379f085fc610f3f2d895c80d5064e
5bd81e4c915b1620e4d190f4b688cf10c53429c5
7f997721693c63afcdf1d23f7449f8acbbe282d1
2f00094498ebd5dc65c66c1219e4553e36ec35f6
fdee485ad02197b9ef020b0a25d09cbb4e639b06
2d6cd1f007c860ad4c76e7efbbb0bf5b4226af0c
df31a989cac681c061051bbd6f53d8a246f41135
8a507796fc37fefdc76cea7a73a3972a7ee945bb
8f8a806fba721f9d363416048e38b7c87e8e0aea
565ce0d3c7ff8634851e91791f7815cb0485ecfb
a7df4ccf0ed9cc594b992dd34b5a740f44e9000f
44d809a90b790dc5d3d9d292d6c7add00e4864b8
a71289508a9783f403ef4a5bd6c5fd7e02a9c42d
32e29e79107a44caa97680c0ff8cad3f06385e13
d6e0094f88e2d80c257388bf88993ccc077f2b5c
55eeb2ae7f56d7ecfa714d661d81e3ca0fb87ebc
5657e7e943f517a252099ccde9c2273530e97584
583965448f2bac1e2f8b6e6fed7aee25bda1fe88
8e401b251f25be5d9daceab12c73b760a771df6e
360aefc01854e9d4bb4163e85adc82354817ecdb
b599cd2acceb4ac96e4d3569bb1caaea4f51f266
6e3af3f17a5f1000822b0a5726920ab4a8c82968
cced3a2d2911bbf9d2adf3b2b47519b9cdc2c0d3
85f5e76175f190ef13b7d883a256612380574ca6
8654a42cd5f96ac6185997e5a3c63f1c69c40a75
3119f6a11455b9c5648167f4c8293476be9c5a9e
37c6708b93af52a4f0e271d217ba973a683ed71f
50cfa27f1b7140edade4b50d94a93c84dee3e267
ae9bfa067e9eae73d4366ddbd8b448823e26eb2d
8dfbc657247d58e9888b7b04b8948558c1218e03
a3267f8b6e939445b016a9dd81e23a63691a10bf
e8b0e4201e8a5793b103c2944d19f7761932a521
f429bd1b24e92433c8cd3c6869953a45aee57b1e
77b76fc900db6db8b6a5f2f5a795123a0be547d7
904ba1c8eea2c01210bf5e5f61680f56c6d3be3d
32e757f4aeb089b764a11f11298c6aab7b1d815d
c4239c05c0e414560361b375fa3606cf0cfcfabb
4b87c90d58a04650d8bc4316115c3233236431a0
51d80884482007f380bd12b4494c893cfd813583
351486633880f1fac39ae10e49b26a1c5e49bc7c
8dbc800f42797105b53fd385eb57ac30bc372d7d
428b4c0a5d798d3a0c1984da7d02e0df86df978e
f44088c33e2984d5dc9e5091ba1452e1aa72d2a8
2614bf87643e1f019fb337fab5fe45c3be5752be
38e40db1c9695786d41a85661e313ce5a207866f
9a1588c4ce9ef163c9f4eeeed1a92e11133ed6c8
a53a11f2182b812250d9470aa4ced536d7fcfb99
815059466cdfb022ff5beec55f7ffee649db7600
8c0a24c14ed9d0ee504463de9d8b8be9da5b5628
602fea614f8272c3fdd9ed7c8a490ad12575425e
bc541126e4058b6139f6b693a1adc9af2332317d
6e72876478bf4dde05ef46b07b15beb818b313b2
bbb322f292cccb0cb60008acb8b50c89b37a8a4a
6384cbcb0eaed417b95e9bce9f666d4f8ef4d7ca
8032b71737520f883abeb1e43476d88cd5a4c304
77a44d94f8d75461dc98fa27a94c8802481ab13e
e4ac313b1280e15c13d12ff3dd25c7c09382920a
12ec584da837e2c3410190b202b7d30875d0a9dc
fcc89a9cfc9fbff3eacc66bdf38ca9dd0f0664f6
3e0372d20e9880c135a8d7ab880cd4105c0222f1
a15e669ceb0c9adde7ca2370aae7168a0b6548bc
9ef49038d59e388a442c265d03d27c4d5e9c7e27
e25f64010ef853728de087ad873e018f34a98382
2171b19c5a9be03429ceb90fb71b95cc9efe1349
9865e1c393ff3c3828a6ff85231c1144e86e03b4
044e54d1eb4cbd009bd2ab267ab03d78c22c2bba
2001be6cc0926bfb5c38d64b74399beff13e00b2
8e7663796904f12ab31320d8d5a056da158483fe
986a6617cc9425dfa4f9fcc879a4d98a3ab00b7c
5f3c9e520c71993ca65a9aaafd9db715eda0b2e5
df9069cd18a3af3d6c6a2452d2daf5a830010bc5
0ad40d16159ff61febd1d884d035dfd8fa47873b
147bad16b96824a3e881650ff89196ddd988ab7f
75f8505d2e63c95a927bca0194ee381930e35592
cba6a409cfea9bf79ffb1088a1a15736cdd4db1a
5be96a2d3a81bea0391722e820c09759da86a0c0
62991a11679da8912815c763980a505d1ad37ed7
73a381871edd64e841ee6040423cfaca25117b1f
7f6077d06322b2ac0d6741c33e644c4c4ec8a95c
60e0eaee230ad2471bf444496241e17305156e45
b7537db6395062ebb8d6ef979eff7a91a6181eec
8d03bc501b5f9cbcb1814d79d750778f56a492da
cc4e065073e430321aeb5c88e925ad9780cf671b
10200d4054e1b2ea9579c5e7b01feee09bf79338
4f096395c15508bb8a8dfc268a2c6103d63988fb
3e347f4181db349299b2dd05bea11b030770af71
e1536fce85b765cdc55993fc1635e030f1952fe2
1909e63acf228bd7e25abaa235e91defa9cc0b63
1b366e63beb7fc9b758c6a61d918c4bf14005bd8
4d21c7c0868e16d37a795dd4da999a1707812c56
e8968ccb2b0b6032b58a2e6a6bedea4766e3cd90
61d9ffe168b4890bb1a91dc2907e00d3c5cd38f2
f820d0250869daea93718402dc337fd1fb4ce7f4
08c51a385dbe501a48536cae74fa3274e08f753c
1ba2db775fe962912b0a089f564ab68d642b0466
7d5bb02b1adf4b0e11c6e27590e0b5b3ab841baa
d1285db253f65a7863d94714bd16925c0e8a52cf
62b045d6c56ef4d1edef9f19228c7e9667c0d6a7
d9f5b2f50f53d625986dbd47cd12778c9f841f0c
07cc1745ff0db626e7e9adc8d18c2372de62a23d
8b5bdc8bdfdb95d5e22ccb8733dbd35c84f3f79d
98fd05580ce702cae1dc403dfd45ae18a6df87cd
d7b5921c580591aca80da561aa84dc32fa4f2222
737170f5bf7e9f999543c30eff4824646885652b
b589493e70323b050c566f465ea3076d5a52b943
9c387098eb147a628857982639258cf7d4d9a090
f0cb8e16f03b37af2f22fe73372fa511dd7104e4
3463f295e79bfc34e4e8164f4faabe2bb6d6ed2a
e9f6516243f58a856bc467a2ca95730cb7cabfc7
a95b8020ca4383dd4a43c18588c004b5683a8566
fec211ad95a3a2967e72d49ab3036ae01e4b3762
d2bec88a36effcff8d91375b325a944dad36b5e4
dd9546e1966e91de0f88bdcba19c93440b8df006
347bb8e26976758a3324c26c661479624eb09080
f17f2c5d6c4321ecac68f55aeba5fbb2d04c27bb
9a0d4dd39be7cbed048e45b5a27ceb81608d76f6
a7df21c4d23ff55f7df3d1a1caabf0040f110edf
da3ef85b22060081cd137a74db42e79d5402fb0e
eecacedc3b9526b59ef690bce41f41158a137c9f
40a2a6cdd3f28848bedd98bfd854c73f788ab086
6b46851168ef37eaacba28a2a655e15ae5934cd0
b2665a211ed7436dbcd4b477f628c2cdf546c2ca
We cannot bisect more!
```

sworleys commented:
Hey @piotrjurkiewicz i will take a look.

Since this patch, zebra handles most things just fine so it might be some special case i missed.

Can you provide your config so I can reproduce?

Im curious as to why your connected routes are inactive.

piotrjurkiewicz commented:
The regressions are associated with ECMP path number limiting using `zebra -e X`.

When `-e X` parameter is not used, everything works fine.

With `-e 1`, all routes are inactive in zebra and they are not pushed down to datapath. This should limit the number of paths to one, not to zero. There is probably off by one error somewhere in the new code.

With `-e 2`, the number of active nexthops is limited to one (instead of two). Zebra `show ip route` command shows only one next hop being active:

```
E>* 10.127.1.252/30 [90/40960] via 10.127.0.102, eth1, 00:04:56
  *                            via 10.127.0.113, eth4 inactive, 00:04:56
  *                            via 10.127.0.125, eth3 inactive, 00:04:56
```

```
  "10.127.1.252\/30":[
    {
      "prefix":"10.127.1.252\/30",
      "protocol":"eigrp",
      "selected":true,
      "destSelected":true,
      "distance":90,
      "metric":40960,
      "installed":true,
      "table":254,
      "internalStatus":16,
      "internalFlags":8,
      "internalNextHopNum":3,
      "internalNextHopActiveNum":1,
      "uptime":"00:02:41",
      "nexthops":[
        {
          "flags":3,
          "fib":true,
          "ip":"10.127.0.102",
          "afi":"ipv4",
          "interfaceIndex":3,
          "interfaceName":"eth1",
          "active":true
        },
        {
          "flags":0,
          "ip":"10.127.0.113",
          "afi":"ipv4",
          "interfaceIndex":4,
          "interfaceName":"eth4"
        },
        {
          "flags":0,
          "ip":"10.127.0.125",
          "afi":"ipv4",
          "interfaceIndex":5,
          "interfaceName":"eth3"
        }
      ]
    }
```

```
ID: 50
     RefCnt: 19
     VRF: default
     Valid, Installed
     Depends: (32) (45) (48)
           10.127.0.102, eth1 (vrf default)
           10.127.0.113, eth4 (vrf default) inactive
           10.127.0.125, eth3 (vrf default) inactive
```

But it seems that all three next hops are installed to kernel (another bug?). Kernel routing table (`ip route`):

```
10.127.1.252/30 nhid 50 proto eigrp metric 20
        nexthop via 10.127.0.125 dev eth3 weight 1
        nexthop via 10.127.0.102 dev eth1 weight 1
        nexthop via 10.127.0.113 dev eth4 weight 1
```

sworleys commented:
@piotrjurkiewicz See #5405 and let me know if it fixes the problem for you.

Thanks for tracking this patch down and reporting the issue directly to me!

piotrjurkiewicz commented:
Yes, #5405 fixes to problems. Thank you.
